x-pack/plugins/alerting/server/alert_instance/create_alert_instance_factory.ts

```diff
@@ -12,12 +12,36 @@ export function createAlertInstanceFactory<
   InstanceState extends AlertInstanceState,
   InstanceContext extends AlertInstanceContext,
   ActionGroupIds extends string
->(alertInstances: Record<string, AlertInstance<InstanceState, InstanceContext, ActionGroupIds>>) {
-  return (id: string): AlertInstance<InstanceState, InstanceContext, ActionGroupIds> => {
-    if (!alertInstances[id]) {
-      alertInstances[id] = new AlertInstance<InstanceState, InstanceContext, ActionGroupIds>();
-    }
+>(
+  alerts: Record<string, AlertInstance<InstanceState, InstanceContext, ActionGroupIds>>,
+  hasRecoveryUtils: boolean,
+  originalAlertIds: Set<string>,
+  getRecoveredAlertIds: (
+    alerts: Record<string, AlertInstance<InstanceState, InstanceContext, ActionGroupIds>>,
+    originalAlertIds: Set<string>
+  ) => string[],
+  recoveryContext: Record<string, InstanceContext>
+) {
+  return {
+    create: (id: string): AlertInstance<InstanceState, InstanceContext, ActionGroupIds> => {
+      if (!alerts[id]) {
+        alerts[id] = new AlertInstance<InstanceState, InstanceContext, ActionGroupIds>();
+      }
 
-    return alertInstances[id];
+      return alerts[id];
+    },
+    done: () => {
+      return hasRecoveryUtils
```

Comment:
I like this approach of returning the recovery APIs only by calling `done()`.  But it seems like we should also prevent other uses of the instance factory (to create new instances), and also prevent existing instances from scheduling actions.  Like, `done()` should set an `isDone` flag that other APIs could check and ... not do what they were going to do.  I think we throw errors for cases like this, but I'd follow whatever we currently do for similar semantic errors with the factory / instances.